pkg/auth/manager/oidcmapping/oidcmapping.go

```diff
@@ -0,0 +1,291 @@
+		am.oidcUsersMapping[u.OIDCGroup] = u
+	}
+
+	return nil
+}
+
+// Authenticate clientID would be empty as we only need to validate the clientSecret variable
+// which contains the access token that we can use to contact the UserInfo endpoint
+// and get the user claims.
+func (am *mgr) Authenticate(ctx context.Context, clientID, clientSecret string) (*user.User, map[string]*authpb.Scope, error) {
+	ctx = am.getOAuthCtx(ctx)
+	log := appctx.GetLogger(ctx)
+
+	oidcProvider, err := am.getOIDCProvider(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("oidcmapping: error creating oidc provider: +%v", err)
+	}
+
+	oauth2Token := &oauth2.Token{
+		AccessToken: clientSecret,
+	}
+
+	// query the oidc provider for user info
+	userInfo, err := oidcProvider.UserInfo(ctx, oauth2.StaticTokenSource(oauth2Token))
+	if err != nil {
+		return nil, nil, fmt.Errorf("oidcmapping: error getting userinfo: +%v", err)
+	}
+
+	// claims contains the standard OIDC claims like issuer, iat, aud, ... and any other non-standard one.
+	// TODO(labkode): make claims configuration dynamic from the config file so we can add arbitrary mappings from claims to user struct.
+	var claims map[string]interface{}
+	if err := userInfo.Claims(&claims); err != nil {
+		return nil, nil, fmt.Errorf("oidcmapping: error unmarshaling userinfo claims: %v", err)
+	}
+
+	log.Debug().Interface("claims", claims).Interface("userInfo", userInfo).Msg("unmarshalled userinfo")
+
+	if claims["issuer"] == nil { // This is not set in simplesamlphp
+		claims["issuer"] = am.c.Issuer
+	}
+	if claims["email_verified"] == nil { // This is not set in simplesamlphp
+		claims["email_verified"] = false
+	}
+	if claims["email"] == nil {
+		return nil, nil, fmt.Errorf("oidcmapping: no \"email\" attribute found in userinfo: maybe the client did not request the oidc \"email\"-scope")
+	}
+	if claims["preferred_username"] == nil || claims["name"] == nil {
+		return nil, nil, fmt.Errorf("oidcmapping: no \"preferred_username\" or \"name\" attribute found in userinfo: maybe the client did not request the oidc \"profile\"-scope")
+	}
+	if claims["groups"] == nil {
+		return nil, nil, fmt.Errorf("oidcmapping: no \"groups\" attribute found in userinfo")
+	}
+
+	// discover the user username
+	var username string
+	mappings := make([]string, 0, len(am.oidcUsersMapping))
+	for _, v := range am.oidcUsersMapping {
+		if v.OIDCIssuer == claims["issuer"] {
+			mappings = append(mappings, v.OIDCGroup)
+		}
+	}
+	intersection := intersect.Simple(claims["groups"], mappings)
+	if len(intersection) > 1 {
+		// multiple mappings is not implemented, we don't know which one to choose
+		return nil, nil, errors.New("oidcmapping: mapping failed, more than one mapping found")
+	}
+	if len(intersection) == 1 {
+		for _, m := range intersection {
+			username = am.oidcUsersMapping[m.(string)].Username
+		}
+	}
+	if username == "" {
+		return nil, nil, errors.New("oidcmapping: unable to retrieve username from mappings")
```

Comment:
That could be the "default" case as provided by the normal oidc driver

Reply:
In de existing oidc driver the user groups are requested from locally, is that correct?

Reply:
That ought to be correct as I saw the current driver is used in production for CERNBox. I guess you are referring to:

https://github.com/cs3org/reva/blob/master/pkg/auth/manager/oidc/oidc.go#L159

Which in our deployment corresponds to a (cached) query to keycloak for the local groups.

Reply:
Is keycloak or indigo the IAM in this deployment?

Reply:
What we're targeting for the mapping is the ESCAPE implementation (which I understand is based on indigo). But the same driver ought to work with keycloak too.